gobblin-data-management/src/main/java/gobblin/data/management/partition/FileSet.java

```diff
@@ -13,38 +13,55 @@
 package gobblin.data.management.partition;
 
 import lombok.AccessLevel;
+import lombok.Builder;
 import lombok.Data;
+import lombok.Getter;
 import lombok.NonNull;
 import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+import lombok.Singular;
 
+import java.io.IOException;
 import java.util.Collection;
 import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
 
+import com.beust.jcommander.internal.Maps;
+import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Lists;
 
 import gobblin.data.management.copy.CopyEntity;
+import gobblin.data.management.copy.CopyableFile;
 import gobblin.dataset.Dataset;
+import gobblin.util.request_allocation.Request;
+import gobblin.util.request_allocation.Requestor;
+
 
 /**
- * A named subset of {@link File}s in a {@link Dataset}. (Useful for partitions, versions, etc.)
+ * A named subset of {@link File}s in a {@link Dataset}. (Useful for partitions, versions, etc.).
+ *
+ * The actual list of files in this {@link FileSet} is, in ideal circumstances, generated lazily. As such, the method
+ * {@link #getFiles()} should only be called when the actual list of files is needed.
  */
-@Data
-@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
-public class FileSet<T extends CopyEntity> {
+@RequiredArgsConstructor(access = AccessLevel.PROTECTED)
+public abstract class FileSet<T extends CopyEntity> implements Request<FileSet<CopyEntity>> {
```

Comment:
I find lazy `Iterable` less intuitive than a method that is explicitly called `generateObject`. There are other things like we still want to extend the class to provide the prioritizer with first-class metadata (for example, the prioritizer can easily access the Hive table information of a `HiveFileSet`), and this class was never used directly previously (constructor was private) so it is not an intrusive change.
Having said that, I don't mind changing to the `Iterable` model.